logicalplan/plan.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"time"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+var DefaultOptimizers = []Optimizer{
+	SortMatchers{},
+	MergeSelectsOptimizer{},
+}
+
+type Plan interface {
+	RunOptimizers([]Optimizer) parser.Expr
+}
+
+type Optimizer interface {
+	Optimize(parser.Expr) parser.Expr
```

Comment:
Perhaps we should also add `Explain()` methods here for sake of debuggability?

Reply:
Yeah I'd love to add this, but maybe in a follow up PR? I am concerned this change might just keep growing.

Reply:
fine to add later

Reply:
It'd be such a killer feature to have `EXPLAIN` for PromQL indeed. Seeing how the physical storage is going to be queried. :100: